src/scripts/clipperUI/ratingsHelper.ts

```diff
@@ -0,0 +1,294 @@
+	 *      a) The major version last seen by the user is greater than the major version at the time of the last bad rating, OR
+	 *      b) The major version remained the same, while the minor version last seen by the user is greater than
+	 *          the minor version at the time of the last bad rating
+	 *
+	 * Public for testing
+	 */
+	public static badRatingVersionDelayIsOver(badRatingVersionAsStr: string, lastSeenVersionAsStr: string): boolean {
+		if (Utils.isNullOrUndefined(badRatingVersionAsStr)) {
+			// value has never been set, no bad rating given
+			return true;
+		}
+
+		let badRatingVersion: Version;
+		let lastSeenVersion: Version;
+		try {
+			badRatingVersion = new Version(badRatingVersionAsStr);
+			lastSeenVersion = new Version(lastSeenVersionAsStr);
+		} catch (e) {
+			return false;
+		}
+
+		return lastSeenVersion.isGreaterThan(badRatingVersion, true /* ignorePatchUpdate */);
+	}
+
+	/**
+	 * Returns true if ALL of the below applies:
+	 *   * Number of successful clips >= {Constants.Settings.minClipSuccessForRatingsPrompt}
+	 *   * Number of successful clips <= {Constants.Settings.maxClipSuccessForRatingsPrompt}
+	 *
+	 * Public for testing
+	 */
+	public static clipSuccessDelayIsOver(numClips: number): boolean {
+		// TODO MVP+? when # of successful clips > nMax, collapse panel into a Rate Us hyperlink in the footer that is always available
+
+		if (isNaN(numClips)) {
+			return false;
+		}
+
+		return numClips >= Constants.Settings.minClipSuccessForRatingsPrompt &&
+			numClips <= Constants.Settings.maxClipSuccessForRatingsPrompt;
+	}
+
+	/**
+	 * Implementation of the logic described in the setShowRatingsPromptState(...) description
+	 */
+	private static shouldShowRatingsPromptInternal(clipperState: ClipperState, event: Log.Event.PromiseEvent, logEventInfo: RatingsLoggingInfo): boolean {
+		if (Utils.isNullOrUndefined(clipperState)) {
+			event.setStatus(Log.Status.Failed);
+			event.setFailureInfo({ error: "Clipper state is null or undefined" });
+			return false;
+		}
+
+		if (!Utils.isNullOrUndefined(clipperState.showRatingsPrompt.get())) {
+			// return cached value in clipper state since it already exists
+			logEventInfo.usedCachedValue = true;
+			return clipperState.showRatingsPrompt.get();
+		}
+
+		let ratingsPromptEnabled: boolean = RatingsHelper.ratingsPromptEnabledForClient(clipperState.clientInfo.clipperType);
+		logEventInfo.ratingsPromptEnabledForClient = ratingsPromptEnabled;
+		if (!ratingsPromptEnabled) {
+			return false;
+		}
+
+		let doNotPromptRatingsStr: string = Clipper.Storage.getCachedValue(Constants.StorageKeys.doNotPromptRatings);
+		let lastBadRatingDateAsStr: string = Clipper.Storage.getCachedValue(Constants.StorageKeys.lastBadRatingDate);
+		let lastBadRatingVersion: string = Clipper.Storage.getCachedValue(Constants.StorageKeys.lastBadRatingVersion);
+		let lastSeenVersion: string = Clipper.Storage.getCachedValue(Constants.StorageKeys.lastSeenVersion);
```

Comment:
This is gonna sound crappy but I am wary of this paradigm of knowing when a value is cached versus when it isn't. I think since you have tested it pretty thoroughly it will work here. The options are:

1. Always get the fresh value from `localStorage`, which means dealing with callbacks, which is why we switched to this design in the first place
3. Risk not having a fresh value. If the caller needs a fresh value, force them to use the async version.

What happens when you attempt `getCachedValue` but a corresponding `getValue` was never called? It is hard to differentiate between not being in storage, as we have a lot of cases where it doesn't exist yet, versus it not being in the cache. From `preCacheNeededValues` I know that these values are fine but we should discuss how we want to treat this in the future. 

TL;DR: `async` makes everything difficult -.-